Add metric properties to the ontology

FransKnibbe opened the pull request:
This adds metric properties to the ontology. It covers [issue 142](https://github.com/opengeospatial/ogc-geosparql/issues/142). 

Note:
- The metric properties do not replace existing properties for size and resolution. That means they can be viewed as convenience properties.
- The PR does include a property for accuracy, which has not been included otherwise yet. I have included it because it was little extra effort.
- ~~If this PR is accepted I can then make a new PR to update the documentation.~~

nicholascar commented:
Before getting into the specifics of this new content (next comment!), can we have a corresponding update to the Spec doc as well? Every ontology property needs its corresponding Spec write up. Surely somewhere in [Section 8.3](https://opengeospatial.github.io/ogc-geosparql/geosparql11/spec.html#_standard_properties_for_geogeometry).

nicholascar commented:
Can the naming of this proposed `hasMetricResolution` be harmonised with the existing `hasSpatialResolution`, i.e. `hasMetricResolution` &rarr; `hasMetricSpatialResolution`?

There seems to be no non-metric equivalent to `hasMetricAccuracy`. Should one be added?

nicholascar commented:
I am in favour of adding these properties as long as equivalence to the existing non-metric ones are shown, likely just through examples.

FransKnibbe commented:
> Before getting into the specifics of this new content (next comment!), can we have a corresponding update to the Spec doc as well? Every ontology property needs its corresponding Spec write up. Surely somewhere in [Section 8.3](https://opengeospatial.github.io/ogc-geosparql/geosparql11/spec.html#_standard_properties_for_geogeometry).
> 
> As a stretch goal: could an example of each of these properties in se be added to [Annex B 1.2](https://opengeospatial.github.io/ogc-geosparql/geosparql11/spec.html#_b_1_2_properties). In particular, could an example show equivalence between direct use of a metric property and the `numericValue` + `unit` examples in [Annex B 1.2.1](https://opengeospatial.github.io/ogc-geosparql/geosparql11/spec.html#_b_1_2_1_feature_properties), just to show that you can either use the metric property or equivalent general property + units of measure.

Ok, I will add both to the PR (I first wanted to be reasonably sure that these additions were acceptable before updating documentation and examples).

FransKnibbe commented:
> There seems to be no non-metric equivalent to `hasMetricAccuracy`. Should one be added?

I could add it, but to be honest the addition was scheduled for version  1.2 (see [issue 23](https://github.com/opengeospatial/ogc-geosparql/issues/23). But since most to the effort goes into the definition of accuracy, which is already proposed for `hasMetricAccuracy`, I'll add the non-metric equivalent too. It would be strange otherwise. Of course people can still reject adding them both....

FransKnibbe commented:
I have updated the PR:

- Added hasSpatialAccuracy similar to hasSpatialResolution
- Added documentation
- Added two examples
- Changed datatype from xsd:decimal to xsd:double to allow scientific notation (otherwise I couldn't get the example to work; hurray for examples!)

I think it is ready for re-review now.

nicholascar reviewed (changes requested):
Hi @FransKnibbe looks comprehensive with examples, Spec wording so pretty good. I fixed a couple of really tiny typos.

Just one final point from me now: starting with GeoSPARQL 1.1 property naming is of the style **has**XXXX, e.g. _hasGeometry_, but for this set of properties you've not included "has". In keeping with the existing style, e.g. _hasArea_, can you rename your additions with "has", e.g. _hasMetricArea_? Thanks.

FransKnibbe commented:
> Just one final point from me now: starting with GeoSPARQL 1.1 property naming is of the style **has**XXXX, e.g. _hasGeometry_, but for this set of properties you've not included "has". In keeping with the existing style, e.g. _hasArea_, can you rename your additions with "has", e.g. _hasMetricArea_? Thanks.

@nicholascar, I removed the ¨has¨ on purpose. I got the impression that there are two different styles in use: object properties with ¨has¨ and datatype properties without ¨has¨. It seemed to make some linguistic sense. For example, you could say that a car has an owner and its weight is 1500 kg. The only datatype property with ¨has¨ that I see now is ¨hasSerialisation¨, which kind of confirms the idea that properties starting with ¨has¨ reference more complex things than single values.

But it was just an impression I had, it doesn´t matter much to me either way.

nicholascar commented:
Looks like not nearly all the ObjectProperties  start with "has", e.g.:

* `ehContains`
* `ehCoveredBy`
* ...
* `sfCrosses`
* ...

My estimate is that "has" was present for `hasGeometry` and we've added `hasArea` etc for the geometry or spatial measure series of properties for a Feature but there was no hard and fast rule for _has_ in GeoSPARQL 1.0.

Clearly we can't alter either the existing `hasGeometre` or the `sf***` & `eh***` properties, so we can't fix on a ObjectProperty (_has_) and a Datatypeproperty (_no has_) approach.

We have _species_ of properties though: 

* spatial relations - `sf***`, `eh***` etc
* Feature/Geometry or SpatialMeasure relations
* datatype properties - `as***` (e.g. `asKML`) and other non-has properties.

In this case of these new properties though, I think they are more similar to the _has_ starting properties they shadow than other datatype properties. i.e. `metricLength` is more similar to `hasLength` than `coordinateDimension` etc, so I'd much prefer all these new properties to start with _has_ too, to match their pairs.

(If I could alter GeoSPARQL 1.0, I would also ad _has_ too all their properties, or remove them all. There is a tend to add _has_ widely (e.g. SOSA & PROF & DCAT) so I'd probably add it, but these options aren't available :( ).

dr-shorthair commented on 1.1/geo.ttl:
```diff
@@ -580,6 +589,53 @@ geosparql:
+:metricSpatialResolution
+	a owl:DatatypeProperty ;
+	rdfs:domain :Geometry ;
+	rdfs:range xsd:double ;
```
I prefer `xsd:decimal` - it allows arbitrary precision and is 'honest' in data transfer.

dr-shorthair commented on 1.1/geo.ttl:
```diff
@@ -580,6 +589,53 @@ geosparql:
+:metricSpatialAccuracy
+	a owl:DatatypeProperty ;
+	rdfs:domain :Geometry ;
+	rdfs:range xsd:double ;
```
I prefer `xsd:decimal` - it allows arbitrary precision and is 'honest' in data transfer

dr-shorthair commented on 1.1/geo.ttl:
```diff
@@ -580,6 +589,53 @@ geosparql:
+	rdfs:isDefinedBy geosparql: , geosparql-spec: ;
+	skos:definition """The spatial resolution of a Geometry in meters."""@en ;
+	skos:note """Spatial resolution specifies the level of detail of a Geometry. It the smallest dinstinguishable distance between spatially adjacent coordinates."""@en;
+	skos:prefLabel "metric spatial resolution"@en ;
```
"spatial resolution in meters" - in English at least 'metric' refers to the whole system of units, not just metres.

dr-shorthair commented on 1.1/geo.ttl:
```diff
@@ -580,6 +589,53 @@ geosparql:
+	rdfs:isDefinedBy geosparql: , geosparql-spec: ;
+	skos:definition """The positional accuracy of the coordinates of a Geometry in meters."""@en ;
+	skos:note """Spatial accuracy is applicable when a Geometry is used to represent a Feature. It is expressed as a distance that indicates the truthfullness of the positions (coordinates) that define the Geometry. In this case accuracy defines a zone surrounding each coordinate within wich the real positions are known to be. The accuracy value defines this zone as a distance from the coordinate(s) ...
+	skos:prefLabel "metric spatial accuracy"@en ;
```
"spatial accuracy in meters" - in English at least 'metric' refers to the whole system of units, not just metres.

dr-shorthair commented on 1.1/geo.ttl:
```diff
@@ -580,6 +589,53 @@ geosparql:
 	skos:prefLabel "spatial dimension"@en ;
 .
 
+:metricSpatialResolution
```
This does not follow the DCAT convention where the property name is `*InMeters`. 
However, element IDs are not as important as labels and definitions.

dr-shorthair commented on 1.1/geo.ttl:
```diff
@@ -580,6 +589,53 @@ geosparql:
+	skos:prefLabel "metric spatial resolution"@en ;
+.
+
+:metricSpatialAccuracy
```
This does not follow the DCAT convention where the property name is `*InMeters`. 
However, element IDs are not as important as labels and definitions.

dr-shorthair commented on 1.1/geo.ttl:
```diff
@@ -580,6 +589,53 @@ geosparql:
+	rdfs:range xsd:double ;
+	rdfs:isDefinedBy geosparql: , geosparql-spec: ;
+	skos:definition """The length of a Feature in meters."""@en ;
+	skos:prefLabel "metric length"@en ;
```
"length in meters"

dr-shorthair commented on 1.1/geo.ttl:
```diff
@@ -580,6 +589,53 @@ geosparql:
+	rdfs:range xsd:double ;
+	rdfs:isDefinedBy geosparql: , geosparql-spec: ;
+	skos:definition """The area of a Feature in square meters."""@en ;
+	skos:prefLabel "metric area"@en ;
```
"area in square meters"

dr-shorthair commented on 1.1/geo.ttl:
```diff
@@ -580,6 +589,53 @@ geosparql:
+	rdfs:range xsd:double ;
+	rdfs:isDefinedBy geosparql: , geosparql-spec: ;
+	skos:definition """The volume of a Feature in cubic meters."""@en ;
+	skos:prefLabel "metric volume"@en ;
```
"volume in cubic meters" - in vernacular English 'metric' denotes the whole system of units, not just metres.

dr-shorthair commented on 1.1/spec/09-Part-06.adoc:
```diff
@@ -191,6 +191,18 @@ TIP: A consistency check can be applied to geometries indicating both this prope
+    skos:prefLabel "metric length"@en ;
+    skos:definition "The length of a Feature in meters."@en ; 
+    rdfs:domain geo:Feature ; 
+    rdfs:range  xsd:double .
```
`xsd:decimal` ?

dr-shorthair commented on 1.1/spec/09-Part-06.adoc:
```diff
@@ -217,6 +229,18 @@ TIP: A consistency check can be applied to geometries indicating both this prope
+    skos:prefLabel "metric area"@en ;
+    skos:definition "The area of a Feature in square meters."@en ; 
+    rdfs:domain geo:Feature ; 
+    rdfs:range  xsd:double .
```
`xsd:decimal` ?

dr-shorthair commented on 1.1/spec/09-Part-06.adoc:
```diff
@@ -244,6 +268,18 @@ TIP: A consistency check can be applied to geometries indicating both this prope
+    skos:prefLabel "metric volume"@en ;
+    skos:definition "The volume of a Feature in cubic meters."@en ; 
+    rdfs:domain geo:Feature ; 
+    rdfs:range  xsd:double .
```
`xsd:decimal` ?

dr-shorthair commented on 1.1/spec/11-Part-08.adoc:
```diff
@@ -111,6 +113,46 @@ geo:hasSpatialResolution a rdf:Property, owl:ObjectProperty;
+    skos:prefLabel "metric spatial resolution"@en ;
+    skos:definition "The spatial resolution of a Geometry in meters."@en ; 
+    rdfs:domain geo:Geometry ;
+    rdfs:range xsd:double .
```
`xsd:decimal` ?

dr-shorthair commented on 1.1/spec/11-Part-08.adoc:
````diff
@@ -111,6 +113,46 @@ geo:hasSpatialResolution a rdf:Property, owl:ObjectProperty;
+```turtle
+geo:metricSpatialResolution a rdf:Property, owl:ObjectProperty;
+    rdfs:isDefinedBy geo: ;
+    skos:prefLabel "metric spatial resolution"@en ;
````
"spatial resolution in meters"

dr-shorthair commented on 1.1/spec/11-Part-08.adoc:
````diff
@@ -111,6 +113,46 @@ geo:hasSpatialResolution a rdf:Property, owl:ObjectProperty;
+```turtle
+geo:metricSpatialAccuracy a rdf:Property, owl:ObjectProperty;
+    rdfs:isDefinedBy geo: ;
+    skos:prefLabel "metrci spatial accuracy"@en ;
````
"spatial accuracy in meters"

dr-shorthair commented on 1.1/spec/11-Part-08.adoc:
```diff
@@ -111,6 +113,46 @@ geo:hasSpatialResolution a rdf:Property, owl:ObjectProperty;
+    skos:prefLabel "metrci spatial accuracy"@en ;
+    skos:definition "The positional accuracy of the coordinates of a Geometry in meters."@en ; 
+    rdfs:domain geo:Geometry ;
+    rdfs:range xsd:double .
```
`xsd:decimal` ?

dr-shorthair commented on 1.1/spec/17-Annex-B.adoc:
```diff
@@ -76,7 +76,21 @@ Here a `Feature` and a `Geometry` are described as per the previous example but
+        geo:hasGeometry [
+            geo:asWKT "MULTIPOLYGON (((149.06016596 -35.23610602, 149.060620714 -35.236043434, ... , 149.06016596 -35.23610602)))"^^geo:wktLiteral ;
+    ] ;
+    geo:metricArea 8.7E4^^xsd:double;
```
87400.

Or `"8.74E4"^^xsd:double`

dr-shorthair commented on 1.1/spec/17-Annex-B.adoc:
```diff
@@ -126,7 +140,25 @@ In this example, `Feature X` has two different `Geometry` instances indicated wi
+eg:x a geo:Feature ;
+    skos:prefLabel "Feature X";
+    geo:hasGeometry [
+        geo:metricSpatialResolution 100^^xsd:float ;
```
`100.`

Or `"100"^^xsd:double`

dr-shorthair commented on 1.1/spec/17-Annex-B.adoc:
```diff
@@ -126,7 +140,25 @@ In this example, `Feature X` has two different `Geometry` instances indicated wi
+        geo:asWKT "MULTIPOLYGON (((149.06016 -35.23610, 149.060620 -35.236043, ... , 149.06016 -35.23610)))"^^geo:wktLiteral ;
+    ] ,
+    [
+        geo:metricSpatialResolution 5^^xsd:float ;
```
`5.`

Or `"5.0"^^xsd:double`

dr-shorthair commented:
1. I suggest a bunch of changes 'metric' to 'in meters" because in vernacular English 'metric' denotes the system of units, not just metre. 
2. I prefer `xsd:decimal` because it is more honest in a serialization. But I know there are different opinions on whether the focus should be on data transfer or in-memory representation

FransKnibbe commented:
@dr-shorthair: thanks for the prompt review.

> 1. I suggest a bunch of changes 'metric' to 'in meters" because in vernacular English 'metric' denotes the system of units, not just metre.

To be clear, you suggest to change only the `skos:prefLabel`s, right? I see not problem with that.

> 2. I prefer `xsd:decimal` because it is more honest in a serialization. But I know there are different opinions on whether the focus should be on data transfer or in-memory representation

I started out with `xsd:decimal`, but then I got into trouble with one of the examples where a value of 8.7 hectometre was used. To express that with  `xsd:decimal` would introduce false precision (8700). So I realized the ability to use scientific notation was needed. I then chose `xsd:double` because it has the highest capability to represent high precision numbers. Floating point numbers have disadvantages, but in my mind they do not outweigh the need for using significant digits.

dr-shorthair commented:
The words in the label (and textual definition) are the most important. 
However, the property-name also matters - I would suggest also making those `AAAInMeters` rather than `metricAAA`. 

I hadn't thought that a lot of trailing zeros gave an impression of a defined precision. Hmmm. There is another issue with double and float described here https://github.com/qudt/qudt-public-repo/issues/307 Perhaps it is not relevant in this slot, but worth a passing thought at least.

FransKnibbe commented:
@dr-shorthair: I have made the following changes:

1. Added "has" to the properties
2. Changed the labels from "metric ...." to ".... in meters"

I haven't changed the property names (yet), with the idea that the current names are shorter and sort of form a group because their names all start with "hasMetric"; it makes them a bit easier to distinguish from their counterparts that can have arbitrary units.

Thanks for pointing out the [QUDT issue](https://github.com/qudt/qudt-public-repo/issues/307). It seems to me that the cases are not fully comparable. The QUDT issue is about conversion factors, while the metric properties in GeoSPARQL will be used for values that are measured or estimated. It is possible to imagine a conversion factor that has too many digits for xsd:double. But for GeoSPARQL we could assume that xsd:double offers enough precision to express measured or estimated values. Not being able to prevent false precision would be a much higher penalty, I think.

Can you accept the PR in its current state?

dr-shorthair commented:
> Thanks for pointing out the QUDT issue. It seems to me that the cases are not fully comparable. 

Yes, I agree that the QUDT issue is addressing a slightly different issue. But thought you would be interested to see a practical analysis of decimal vs. double ;-)

FransKnibbe commented:
@nicholascar: Could you please take a look at this PR again to see if it can be approved?
